pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -118,31 +120,147 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, upd
 		servicePorts[string(svcHost)][port] = cluster
 	}
 
-	// In delta, we only care about the services that have changed.
 	for key := range updates.ConfigsUpdated {
-		// get the service that has changed.
-		service := updates.Push.ServiceForHostname(proxy, host.Name(key.Name))
-		// if this service removed, we can conclude that it is a removed cluster.
-		if service == nil {
-			for cluster := range serviceClusters[key.Name] {
-				deletedClusters = append(deletedClusters, cluster)
+		switch key.Kind {
+		case kind.ServiceEntry:
+			services, deletedClusters = configgen.deltaFromServices(key, proxy, updates.Push, watched.ResourceNames, serviceClusters, servicePorts)
+		case kind.DestinationRule:
+			services, deletedClusters = configgen.deltaFromDestinationRules(key, proxy, updates.Push, watched.ResourceNames)
+		}
+	}
+	clusters, log := configgen.buildClusters(proxy, updates, services)
+	return clusters, deletedClusters, log, true
+}
+
+// deltaFromServices computes the delta clusters from the updated services.
+func (configgen *ConfigGeneratorImpl) deltaFromServices(updatedService model.ConfigKey, proxy *model.Proxy, push *model.PushContext,
+	watched []string, serviceClusters map[string]sets.String, servicePorts map[string]map[int]string,
+) ([]*model.Service, []string) {
+	var deletedClusters []string
+	var services []*model.Service
+	service := push.ServiceForHostname(proxy, host.Name(updatedService.Name))
+	// push.ServiceForHostname will return nil if the proxy doesn't care about the service OR it was deleted.
+	// we can cross-reference with WatchedResources to figure out which services were deleted.
+	if service == nil {
+		// We assume a service was deleted, it doesn't hurt to have positives for removedResources when a resource included
+		// does not exist.
+		// WatchedResources.ResourceNames will contain the names of the clusters it is subscribed to. We can
+		// check with the name of our service (cluster names are in the format outbound|<port>||<hostname>
+		// so, we can check if the cluster names contains the service name, and determine if it is deleted.
+		deletedClusters = append(deletedClusters, removedClusterNames(watched, nil, updatedService.Name)...)
```

Comment:
serviceClusters already contains something similar to removedClusterNames  and is not used. We should remove serviceClusters or use it (probably use it, since it pre-processes outside the loop of ConfigsUpdated)